### PR TITLE
res_prometheus: Fix duplicate output of metric and help text

### PR DESCRIPTION
**Pull request by zecke**
The prometheus exposition format requires each line to be unique[1]. This is handled by struct prometheus_metric having a list of children that is managed when registering a metric. In case the scrape callback is used, it is the responsibility of the implementation to handle this correctly.

Originally the bridge callback didn't handle NULL snapshots, the crash fix lead to NULL metrics, and fixing that lead to duplicates.

The original code assumed that snapshots are not NULL and then relied on "if (i > 0)" to establish the parent/children relationship between metrics of the same class. This is not workerable as the first bridge might be invisible/lacks a snapshot.

Fix this by keeping a separate array of the first metric by class. Instead of relying on the index of the bridge, check whether the array has an entry. Use that array for the output.

Add a test case that verifies that the help text is not duplicated.

Resolves: #642

[1] https://prometheus.io/docs/instrumenting/exposition_formats/#grouping-and-sorting

### CODE REVIEW

File: res/prometheus/bridges.c [hunk 1]
```diff
@@ -116,12 +117,18 @@ static void bridges_scrape_cb(struct ast_str **response)
...
+	if (!metrics) {
 		ao2_ref(bridges, -1);
 		return;
 	}
 
+	bridge_metrics = ast_calloc(ARRAY_LEN(bridge_metric_defs), sizeof(bridge_metrics));
+	if (!bridge_metrics) {
+		ao2_ref(bridges, -1);
+		ast_free(metrics);
+	}
```
Comment (seanbright) [inline res/prometheus/bridges.c:130]: `return;` here?

File: tests/test_res_prometheus.c [hunk 2]
```diff
@@ -710,10 +710,23 @@ static void safe_bridge_destroy(struct ast_bridge *bridge)
 	ast_bridge_destroy(bridge, 0);
 }
 
+static int word_count(const char *str, const char *needle)
```
Comment (seanbright) [inline tests/test_res_prometheus.c:713]: Minor: `word_count` sounds like it counts words. `match_count` instead?

File: res/prometheus/bridges.c [hunk 3]
```diff
@@ -116,12 +117,18 @@ static void bridges_scrape_cb(struct ast_str **response)
...
+	metrics = ast_calloc(ARRAY_LEN(bridge_metric_defs) * num_bridges, sizeof(*metrics));
+	if (!metrics) {
 		ao2_ref(bridges, -1);
 		return;
 	}
 
+	bridge_metrics = ast_calloc(ARRAY_LEN(bridge_metric_defs), sizeof(bridge_metrics));
+	if (!bridge_metrics) {
+		ao2_ref(bridges, -1);
+		ast_free(metrics);
```
Comment (seanbright) [inline res/prometheus/bridges.c:129]: Minor: Maybe swap these two lines to free in the reverse order that we allocated?

**Review by seanbright — CHANGES_REQUESTED** [review]

File: res/prometheus/bridges.c (hunk 3 above)
Comment (zecke) [reply res/prometheus/bridges.c]: Done. I also applied this order at the end of the function.

**Review by gtjoseph — CHANGES_REQUESTED** [review]
@zecke Can you open a "bug" issue for this and add the appropriate "Resolves:" trailer to the commit message and the PR description?

**Review by jcolp — APPROVED** [review]

**Comment by zecke** [comment]
> @zecke Can you open a "bug" issue for this and add the appropriate "Resolves:" trailer to the commit message and the PR description?

Done. Mind having a look?

**Review by gtjoseph — APPROVED** [review]

**Comment by gtjoseph** [comment]
I'm closing and re-opening just to trigger a retest.

**Review by mbradeen — APPROVED** [review]

**Pull request merged**